src/App.js

```diff
@@ -30,36 +30,111 @@ const App = () => {
+
   // Wave 2
   // You will need to create a method to change the square
   //   When it is clicked on.
   //   Then pass it into the squares as a callback
+  const onClickCallback = (squareID) => {
+    if (winner){
+      return;
+    }
+
+    let madeMove = false;
+
+    const newSquares = squares.map(row => row.map(position => {
+      if (position.id != squareID){
+        return position;
+      }
+      if (position.value !== ''){
+        return position;
+      }
+      madeMove = true;
+      const newSquare = {...position, value: currentPlayer};
+      return newSquare;
+    }));
+
+    if (madeMove) {
+      //update the board
+      setSquares(newSquares);
+    
+      // change the player with each click
+      let nextPlayer = (currentPlayer === PLAYER_1) ? PLAYER_2 : PLAYER_1;
+      setCurrentPlayer(nextPlayer);
+    }
+  };
 
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if
-    //    all three squares have the same value.
+    let i = 0;
+
+    // Check all the rows and columns for a winner
+    while (i < 3) {
+      if (
+        squares[i][0].value === squares[i][1].value &&
+        squares[i][2].value === squares[i][1].value &&
+        squares[i][0].value !== ''
+      ) {
+        return squares[i][0].value;
+      } else if (
+        squares[0][i].value === squares[1][i].value &&
+        squares[2][i].value === squares[1][i].value &&
+        squares[0][i].value !== ''
+      ) {
+        return squares[0][i].value;
+      }
+      i += 1;
+    }
+    // Check Top-Left to bottom-right diagonal
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][0].value;
+    }
+
+    // Check Top-right to bottom-left diagonal
+    if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][2].value;
+    }
+
+    return null;
   };
 
   const resetGame = () => {
     // Complete in Wave 4
+    setSquares(generateSquares());
+    setCurrentPlayer(PLAYER_1);
+  };
+
+  const winner = checkForWinner();
+
+  const showStatus = () => {
+    if (winner){
+      // alert(`You did amazing ${winner}!`)
+      return `Winner is ${winner}`;
+    }else{
+      return `Your turn ${currentPlayer}`;
+    }
   };
 
   return (
-    <div className="App">
-      <header className="App-header">
+    <div className='App'>
+      <header className='App-header'>
         <h1>React Tic Tac Toe</h1>
-        <h2>The winner is ... -- Fill in for wave 3 </h2>
-        <button>Reset Game</button>
+        <h2> {showStatus()} </h2>
```

Comment:
Nice helper function! Another approach to displaying winner text is by using a ternary operator:

```suggestion
        <h2> { winner ? Winner is ${winner} : `Your turn ${currentPlayer}` } </h2>
```